pkg/filesystem/path/resolve.go

```diff
@@ -1,133 +1,64 @@
 package path
 
-import (
-	"strings"
-
-	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/status"
-)
-
-func stripOneOrMoreSlashes(p string) string {
-	for {
-		p = p[1:]
-		if p == "" || p[0] != '/' {
-			return p
-		}
-	}
-}
-
 type resolverState struct {
-	stack           []string
+	stack           []RelativeParser
 	componentWalker ComponentWalker
 }
 
-// Push a new path string onto the stack of paths that need to be
+// Push a new path onto the stack of paths that need to be
 // processed. This happens once when the resolution process starts, and
 // will happen for every symlink encountered.
-func (rs *resolverState) push(scopeWalker ScopeWalker, path string) error {
-	// Unix-style paths are generally passed to system calls that
-	// accept C strings. There is no way these can accept null
-	// bytes.
-	if strings.ContainsRune(path, '\x00') {
-		return status.Error(codes.InvalidArgument, "Path contains a null byte")
-	}
-
-	// Determine whether the path is absolute.
-	absolute := false
-	if path != "" && path[0] == '/' {
-		path = stripOneOrMoreSlashes(path)
-		absolute = true
-	}
-
+func (rs *resolverState) push(scopeWalker ScopeWalker, parser Parser) error {
 	// Push the path without any leading slashes onto the stack, so
-	// that its components may be processed. Apply  them against the
+	// that its components may be processed. Apply them against the
 	// right directory.
-	var componentWalker ComponentWalker
-	var err error
-	if absolute {
-		componentWalker, err = scopeWalker.OnAbsolute()
-	} else {
-		componentWalker, err = scopeWalker.OnRelative()
-	}
+	componentWalker, remainder, err := parser.ParseScope(scopeWalker)
 	if err != nil {
 		return err
 	}
-	rs.stack = append(rs.stack, path)
+	if remainder != nil {
```

Comment:
Maybe we can require that Parser.ParseScope() always returns a RelativePath? That way this code can just push it on the stack unconditionally.

Reply:
Yeah, in the absence of an error there will always be a remainder. But the two nilable return vaules do not reflect this. As the code is written it is safe to drop the check, if that is desireable (I prefer the belt-and-suspenders). 
Or I could refactor the ParseScope function to just return everything with the named output arguments,
and then the error being there is all that matters. 

Something something sum-types